### PR TITLE
fix: bump MSRV to 1.71.1

### LINKED ISSUE #1101: imp(ibc-primitives): use `let-else` over `downcast!()` from Rust 1.65+
## Feature Summary

We should use `let-else` over `downcast!()` when we[ support Rust 1.65+](https://blog.rust-lang.org/2022/11/03/Rust-1.65.0.html#let-else-statements).

## Proposal

currently,
```rs
let tm_block = downcast!(block.clone() => HostBlock::SyntheticTendermint).unwrap();
```

after 1.65+ support,
```rs
let HostBlock::SyntheticTendermint(tm_block) = block.clone() else { panic!() };
```

### LINKED ISSUE #1118: bump msrv for the project
## Bug Summary

Currently the MSRV is specified to be 1.64. But it's actually 1.71.1.

## Details

This needs to be fixed in `rust-version` in `Cargo.toml`. We should add a check on CI for different OS.
Also, update the details and badges on READMEs.

## Version

< 0.50.0

### LINKED ISSUE #1120: imp(ibc-core): shadowing with the same value
## Bug Summary

The following two cases are caught by [`clippy::shadow_unrelated`](https://rust-lang.github.io/rust-clippy/master/index.html#/shadow_unrelated). There are other cases, but these two cases are special as they are using the exact same value. So we can remove the shadowing and use the first binding directly.

- ibc-core/ics04-channel/src/handler/acknowledgement.rs

https://github.com/cosmos/ibc-rs/blob/e1f19af161878f9daf779aed0f19229f97d54814/ibc-core/ics04-channel/src/handler/acknowledgement.rs#L55-L59

is shadowed by

https://github.com/cosmos/ibc-rs/blob/e1f19af161878f9daf779aed0f19229f97d54814/ibc-core/ics04-channel/src/handler/acknowledgement.rs#L77-L81

Since `msg` is immutable, both are the same. We can remove second one and use the previous variable directly. 

- ibc-core/ics04-channel/src/handler/timeout.rs

https://github.com/cosmos/ibc-rs/blob/e1f19af161878f9daf779aed0f19229f97d54814/ibc-core/ics04-channel/src/handler/timeout.rs#L67-L68

is shadowed by

https://github.com/cosmos/ibc-rs/blob/e1f19af161878f9daf779aed0f19229f97d54814/ibc-core/ics04-channel/src/handler/timeout.rs#L85-L89

Since `packet` is immutable, we can use previous one directly.

## Version

<= 0.50.0

### PR DESCRIPTION
**Pull request by rnbguy**
Closes: #1118
Closes: #1101
Closes: #1120

## Description

- Increase MSRV to 1.71.1
- Check MSRV on CI for linux and darwin
- Update README with new links
- Apply some clippy suggestions
______

### PR author checklist:
- [x] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [x] Added tests.
- [x] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [x] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -157,3 +157,48 @@ jobs:
         with:
           command: test
           args: --no-default-features --no-fail-fast --no-run
+
+  test-msrv:
```
Comment (Farhad-Shabani) [inline .github/workflows/rust.yml:161]: Great Job here for the `msrv`. Awesome!

File: ibc-primitives/src/utils/macros.rs [hunk 2]
```diff
@@ -1,51 +0,0 @@
-//! Macros used across the ibc-rs crates.
```
Comment (Farhad-Shabani) [inline ibc-primitives/src/utils/macros.rs:1]: Nice clean-ups!

File: ibc-testkit/src/testapp/ibc/core/types.rs [hunk 3]
```diff
@@ -139,7 +139,7 @@ pub struct MockClientConfig {
     trusting_period: Duration,
     #[builder(default = Duration::from_millis(3000))]
     max_clock_drift: Duration,
-    #[builder(default = Duration::from_secs(128000))]
+    #[builder(default = Duration::from_secs(128_000))]
```
Comment (Farhad-Shabani) [inline ibc-testkit/src/testapp/ibc/core/types.rs:142]: I liked the tidy outcome from `clippy`, however, couldn’t catch them on my machine. What version of `clippy` do you use? Did the CI detect these cases?

As for the particular case of the `128_000`, it's strange that `64000` would be considered acceptable.

File: ibc-core/ics04-channel/src/handler/acknowledgement.rs [hunk 4]
```diff
@@ -74,11 +74,6 @@ where
 
     // apply state changes
     {
-        let commitment_path_on_a = CommitmentPath {
```
Comment (Farhad-Shabani) [inline ibc-core/ics04-channel/src/handler/acknowledgement.rs:77]: Are these path removals as additional clean-ups to this PR?
Would you open up an issue and include it in the changelog?

File: ibc-primitives/src/utils/macros.rs (hunk 2 above)
Comment (rnbguy) [reply ibc-primitives/src/utils/macros.rs:1]: Before this wasn't possible as `let PATTERN = value else { ... }` is stabilized in 1.65 :slightly_smiling_face:

File: ibc-testkit/src/testapp/ibc/core/types.rs (hunk 3 above)
Comment (rnbguy) [reply ibc-testkit/src/testapp/ibc/core/types.rs:142]: I cherry picked the lints after running all clippy lint groups. We shouldn't do this directly on CI (but probably locally for reviewing PRs is fine) - as most of them are very opinionated. But we can definitely add a bunch of them on CI.

Comment (rnbguy) [reply ibc-testkit/src/testapp/ibc/core/types.rs:142]: For the particular case, the lint case is called [`unreadable_literal`](https://rust-lang.github.io/rust-clippy/master/index.html#/unreadable_literal). Looks like they warn this only when the decimal has 6 digits i.e. million.

File: .changelog/unreleased/improvements/1101-use-let-else-over-downcast.md [hunk 5]
```diff
@@ -0,0 +1,2 @@
+- [ibc-primitives] use `let-else` over `downcast!()` and remove `utils/macros`
+  module as a result ([\#1118](https://github.com/cosmos/ibc-rs/issues/1118))
```
Comment (rnbguy) [inline .changelog/unreleased/improvements/1101-use-let-else-over-downcast.md]: Ah thanks for this. I forgot about it.

File: ibc-core/ics04-channel/src/handler/acknowledgement.rs (hunk 4 above)
Comment (rnbguy) [reply ibc-core/ics04-channel/src/handler/acknowledgement.rs:77]: This came up from [`shadow_reuse`](https://rust-lang.github.io/rust-clippy/master/index.html#/shadow_reuse) or [`shadow_unrelated`](https://rust-lang.github.io/rust-clippy/master/index.html#/shadow_unrelated). I didn't add others - but this looked reasonable to be fixed.

I will add an issue.

**Comment by rnbguy** [comment]
btw, I want to fix one more lint but the changes are too big. You can try it by running

```sh
cargo clippy --fix -- -W clippy::use_self
```

File: ibc-core/ics04-channel/src/handler/acknowledgement.rs (hunk 4 above)
Comment (rnbguy) [reply ibc-core/ics04-channel/src/handler/acknowledgement.rs:77]: 332c5de

**Review by Farhad-Shabani — APPROVED** [review]

File: ibc-testkit/src/testapp/ibc/core/types.rs (hunk 3 above)
Comment (Farhad-Shabani) [reply ibc-testkit/src/testapp/ibc/core/types.rs:142]: Let's fix things when the linter catches them. That's alright.

File: ibc-core/ics04-channel/src/handler/acknowledgement.rs (hunk 4 above)
Comment (Farhad-Shabani) [reply ibc-core/ics04-channel/src/handler/acknowledgement.rs:77]: Thank you for the issue 👍🏻

**Pull request merged**